### PR TITLE
[Semi-Auto]Add master_grad for semi-auto parallel in AMP

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Breaking changes
### PR changes

Others
### Description

[Distributed]Add utest for semi-auto parallel
Pcard-73145

### CODE REVIEW

File: test/auto_parallel/semi_auto_parallel_simple_net_master_grad.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    DemoNet,
+    TestSimpleNetForSemiAutoParallel,
+)
+
+import paddle
+import paddle.distributed as dist
+from paddle import nn
+
+
+class TestSimpleNetWithAmpForSemiAutoParallel(TestSimpleNetForSemiAutoParallel):
```
Comment (chenwhql) [inline test/auto_parallel/semi_auto_parallel_simple_net_master_grad.py]: 和amp那个单测重名，且内容重合度也比较高，如果主要是master_grad参数的变化，要不要整理复用一下代码

File: test/amp/test_amp_master_grad.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def check_results(
         )
         self.assertEqual(
             int(op_list['transfer_dtype'].split(',')[0]),
-            total_steps + total_steps * 2,
+            total_steps + total_steps * 2 + 2,  # add 2 for adam's master grad
```
Comment (chenwhql) [inline test/amp/test_amp_master_grad.py]: 这里+2要不要再多写一些注释

Comment (ForFishes) [reply test/amp/test_amp_master_grad.py]: ok, i fill fix it.

File: test/auto_parallel/semi_auto_parallel_simple_net_master_grad.py (hunk 1 above)
Comment (ForFishes) [reply test/auto_parallel/semi_auto_parallel_simple_net_master_grad.py]: i will merge them together.

**Review by XieYunshen — APPROVED** [review]
LGTM for `  set_tests_properties(test_semi_auto_parallel_in_framework
                       PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE" TIMEOUT 300)`

**Review by LiYuRio — APPROVED** [review]

**Pull request merged**